src/snowflake/connector/aio/_file_transfer_agent.py

```diff
@@ -0,0 +1,287 @@
+                    done_client.delete_client_data()
+                else:
+                    await done_client.finish_download()
+            elif done_client.meta.result_status == ResultStatus.SKIPPED:
+                # this case applies to upload only
+                return
+            else:
+                logger.debug(f"Finished preparing file {done_client.meta.name}")
+                tasks = []
+                for _chunk_id in range(done_client.num_of_chunks):
+                    task = (
+                        asyncio.create_task(done_client.upload_chunk(_chunk_id))
+                        if is_upload
+                        else asyncio.create_task(done_client.download_chunk(_chunk_id))
+                    )
+                    task.add_done_callback(
+                        lambda t, ta=task, dc=done_client, _chunk_id=_chunk_id: transfer_done_cb(
+                            ta, dc, _chunk_id
+                        )
+                    )
+                    tasks.append(task)
+                await asyncio.gather(*tasks)
+                await asyncio.gather(*finish_download_upload_tasks)
+
+        def transfer_done_cb(
+            task: asyncio.Task,
+            done_client: SnowflakeStorageClient,
+            chunk_id: int,
+        ) -> None:
+            # Note: chunk_id is 0 based while num_of_chunks is count
+            logger.debug(
+                f"Chunk {chunk_id}/{done_client.num_of_chunks} of file {done_client.meta.name} reached callback"
+            )
+            if task.exception():
+                done_client.failed_transfers += 1
+                logger.debug(
+                    f"Chunk {chunk_id} of file {done_client.meta.name} failed to transfer for unexpected exception {task.exception()}"
+                )
+            else:
+                done_client.successful_transfers += 1
+            logger.debug(
+                f"Chunk progress: {done_client.meta.name}: completed: {done_client.successful_transfers} failed: {done_client.failed_transfers} total: {done_client.num_of_chunks}"
+            )
+            if (
+                done_client.successful_transfers + done_client.failed_transfers
+                == done_client.num_of_chunks
+            ):
+                if is_upload:
+                    finish_upload_task = asyncio.create_task(
+                        done_client.finish_upload()
+                    )
+                    finish_download_upload_tasks.append(finish_upload_task)
+                    done_client.delete_client_data()
+                else:
+                    finish_download_task = asyncio.create_task(
+                        done_client.finish_download()
+                    )
+                    finish_download_task.add_done_callback(
+                        lambda t, task=finish_download_task, dc=done_client: postprocess_done_cb(
+                            task, dc
+                        )
+                    )
```

Comment:
same comment as above:
```suggestion
                    finish_download_task.add_done_callback(
                        lambda t, dc=done_client: postprocess_done_cb(
                            t, dc
                        )
                    )
```